scapy/layers/tls/handshake.py

```diff
@@ -1286,6 +1305,18 @@ def build(self, *args, **kargs):
             self.exchkeys = cls
         return _TLSHandshake.build(self, *args, **kargs)
 
+    def tls_session_update(self, msg_str):
+        """
+        Finalize the EXTMS messages and compute the hash
+        """
+        super(TLSClientKeyExchange, self).tls_session_update(msg_str)
+
+        if self.tls_session.extms:
+            hash_object = self.tls_session.pwcs.hash
```

Comment:
Is it correctly fallback to `MD5+SHA1` for `TLS 1.0` - `TLS 1.1` (as said chapter 3 of RFC 7627)?

I've got traffic examples for `Extended-Master-Secret` for TLS 1.0 & TLS 1.1 only with `Encrypt-then-MAC`.  
So can't check it right now.

Reply:
This should be alright, it's the same hash function as the one defined in PRF over here https://github.com/secdev/scapy/blob/54da22f91ce5fd45d30efd0359305f4d1fa448a9/scapy/layers/tls/crypto/prf.py#L178

Reply:
I'm not a 100% sure, will check that tonight. Should be defined over https://github.com/secdev/scapy/blob/54da22f91ce5fd45d30efd0359305f4d1fa448a9/scapy/layers/tls/session.py#L36 but this is usually pretty alright. This fallback is implemented for some parts like the PRF

Reply:
Checked. Current implementation works only for `TLS 1.2`.  
Here fix: https://github.com/strayge/scapy/commit/b6a97dccb880c0285c893a644fc3e57a74f89192

Also for testing I wrote POC of `Encrypt-Then-MAC`: https://github.com/strayge/scapy/commit/aee5f5712f287256a94ea34a46f577b9434374ae

with both this commits successfully decrypted: `TLS 1.0` - `TLS 1.2` with `EMS` + `ETM` extensions (checked on `AES-CBC` & `AES-GCM` suites)

Reply:
Thanks a lot for the PoC ! Looking good.
There are a few things that require changes (I'll go over that tommorow), such as the case where the server doesn't support ExtMS or EncryptThenMac, despite the client request, but it's a solid start.

Reply:
Extensions detection placed in `TLSServerHello` class, so we detect it only if server support it (and client requested it).

Also in PoC missed implementation of `Encrypt-Then-MAC` for stream ciphers (for AEAD ciphers server should not send this extension, according to RFC).